src/main/java/com/rocketden/main/service/RoomService.java

```diff
@@ -298,4 +299,25 @@ public RoomDto updateRoomSettings(String roomId, UpdateSettingsRequest request)
         socketService.sendSocketUpdate(roomDto);
         return roomDto;
     }
+
+    public RoomDto setSpectator(String roomId, Boolean isSpectator, SetSpectatorRequest request) {
+        Room room = repository.findRoomByRoomId(roomId);
+
+        // Return error if room could not be found
+        if (room == null) {
+            throw new ApiException(RoomError.NOT_FOUND);
+        }
+
+        // Return error if the initiator is not the host or the same as the receiver
+        User initiator = UserMapper.toEntity(request.getInitiator());
+        User receiver = UserMapper.toEntity(request.getReceiver());
+        if (!room.getHost().equals(initiator) && !initiator.equals(receiver)) {
```

Comment:
The `initiator` and `receiver` could potentially both be bogus objects and still pass this, as there is a simple equality check; you have to check that they are real users in the database (or, if we stick with this room incorporation in this PR, check that both users exist in the room).